### PR TITLE
Generate contributors reports based on git history

### PR DESCRIPTION
**Pull request by michaelsembwever**
Examples deployed at https://nightlies.apache.org/cassandra/devbranch/misc/contribulyze/html/

### CODE REVIEW

File: contribulyze/contribulyze.py [hunk 1]
```diff
@@ -0,0 +1,661 @@
...
+            if m:
+                reviewers = re.split(',|&|( |\n)and( |\n)(by( |\n))?', m.group(1))
+                for reviewer in reviewers:
+                    if reviewer and not reviewer.isspace():
+                        c = Contributor.get(" ".join(reviewer.strip().split()), None)
+                        review_field.add_contributor(c)
+                        c.add_activity(review_field, log)
+                log.add_field(review_field)
+            for c in patch_field.contributors:
+                c.add_collaboration(review_field)
```
Comment (michaelsembwever) [inline contribulyze/contribulyze.py:501]:
```suggestion
                c.add_collaboration(patch_field)
                c.add_collaboration(review_field)
```

File: contribulyze/contribulyze.py [hunk 2]
```diff
@@ -0,0 +1,661 @@
...
+            for c in review_field.contributors:
+                c.add_collaboration(patch_field)
```
Comment (michaelsembwever) [inline contribulyze/contribulyze.py:503]:
```suggestion
                c.add_collaboration(patch_field)
                c.add_collaboration(review_field)
```

File: contribulyze/contribulyze.py [hunk 3]
```diff
@@ -0,0 +1,663 @@
...
+            c.add_aliases(name)
+        else:
+            c = Contributor.get(name, None)
+            c.add_aliases(committer)
+        c.is_committer = True
+
+
+### Regexps to parse the logs. ##
+log_header_re = re.compile('^commit ([0-9a-z]+)$')
+patch_by_re = re.compile('(?:.*\n)*.*patch by ([^;]+)(;|,)', flags=re.IGNORECASE | re.MULTILINE)
```
Comment (azotcsit) [inline contribulyze/contribulyze.py:454]: @michaelsembwever 
I feel there is an issue here. Reviewers are parsed as contributors when there is a comma at the end of contributor names and a comma in the list of reviewers, for example: "patch by Berenguer Blasi, reviewed by Tomasz Lasica, Andres de la Peña and Ekaterina Dimitrova". Basically everything until the last comma is considered to be a contributors list. As a result the script creates wrong records in the output: https://nightlies.apache.org/cassandra/devbranch/misc/contribulyze/html/all_time/detail/reviewed%2520by%2520Tomasz%2520Lasica.html (there are more like this).

I think we need to stop when we see start of the reviewers section. Smth like this:
```
patch_by_re = re.compile('(?:.*\n)*.*patch by ([^;]+)[;, ](?:review|test)(?:ed)? by', flags=re.IGNORECASE | re.MULTILINE)
```

I briefly tested it on a few examples and it seems to be working fine, but I had not put a lot of effort into the testing process.

Comment (michaelsembwever) [reply contribulyze/contribulyze.py:454]: thanks @alex-ninja , let me try it out in an extra commit on the branch…

Comment (michaelsembwever) [reply contribulyze/contribulyze.py:454]: [that](https://github.com/apache/cassandra-builds/pull/54/commits/ab03d1758a9d5e11aad45ba1fcc14ab09c35edb3) didn't work.

Example results:

Screenshot 2021-09-17 at 18 44 15

Screenshot 2021-09-17 at 18 45 05

lots of missing "patches" in those lists. removing that commit for now…

File: contribulyze/contribulyze.py [hunk 4]
```diff
@@ -0,0 +1,663 @@
...
+      c.add_aliases(alias.strip())
+    line = aliases_input.readline()
+
+def process_committers():
+    committers_url = 'https://whimsy.apache.org/public/public_ldap_projects.jsonn'
+    names_url = 'https://whimsy.apache.org/public/icla-info.json'
+    committers = json.loads(requests.get(committers_url).text)['projects']['cassandra']['members']
+    names_json = json.loads(requests.get(names_url).text)['committers']
+    for committer in committers:
+        name = names_json[committer]
```
Comment (JeremiahDJordan) [inline contribulyze/contribulyze.py]: If the `committer` alias isn't in the names list, it should keep that as the name to be looked up.
```suggestion
        name = names_json.get(committer, committer)
```

File: contribulyze/contribulyze.aliases [hunk 5]
```diff
@@ -0,0 +1,69 @@
+Aaron Morton,amorton
+Adam Holmberg,aholmber,Adam Holberg
```
Comment (JeremiahDJordan) [inline contribulyze/contribulyze.aliases:2]:
```suggestion
Adam Holmberg,aholmber,Adam Holberg,reviewed by Adam Holmberg
```

File: contribulyze/contribulyze.aliases [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+Adrian Cole,Adrain Cole
+Aleksandr Sorokoumov,Aleksandr Soromoukov
+Aleksey Yeschenko,iamaleksey,Aleksey,ayeschenko,Aleksey Yeshchenko
+Aleksei Zotov,Alexey Zotov,Aleksei Zoto
```
Comment (JeremiahDJordan) [inline contribulyze/contribulyze.aliases]:
```suggestion
Aleksey Yeschenko,iamaleksey,Aleksey,ayeschenko,Aleksey Yeshchenko,iamaleskey
Aleksei Zotov,Alexey Zotov,Aleksei Zoto,azotcsit
```

File: contribulyze/contribulyze.aliases [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+Andrés de la Peña,Andres de la Peña,adelapena,Andres de la Pena,adelpena
+Ariel Weisberg,aweisberg,Ariel,awiesberg
+Ben Coverston,bcoverston
+Benedict Elliott Smith,Benedict,bes,Benedict Elliot Smith,belliottsmith,belliotsmith,Benedict Elliott-Smith
+Benjamin Lerer,blerer,Benjamin
+Berenguer Blasi,Bereng,Berenguer Blas,Bereguer Blasi
+Blake Eggleston,blake,Blake Egglestone,bdeggleston
+Brandon Williams,brandonwilliams,driftx,brandon.williams,brandwonwilliams,brandonwillams,bwilliams,Brandon Williamms,Brandom Williams,brandonwillias
```
Comment (JeremiahDJordan) [inline contribulyze/contribulyze.aliases:14]:
```suggestion
Brandon Williams,brandonwilliams,driftx,brandon.williams,brandwonwilliams,brandonwillams,bwilliams,Brandon Williamms,Brandom Williams,brandonwillias,reviewed by brandonwilliams
```

File: contribulyze/contribulyze.aliases [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+Branimir Lambov,blambov,branimir
+Carl Yeksigian,carlyeks,carl,cyeksigian
+Chris Goffinet,goffinet
+Chris Lohfink,clohfink,CHris Lohfink
+Dave Brosius,dbrosius
+David Capwell,dcapwell
+Deepak Vohra,dvohra
+Dinesh Joshi,Dinesh A. Joshi
+Eduard Tudenhöfner,Eduard Tudenhoefner
+Ekaterina Dimitrova,Ekaterina Dimotrova,edimitrova,Ekaterina Dimitrov
```
Comment (JeremiahDJordan) [inline contribulyze/contribulyze.aliases:24]:
```suggestion
Ekaterina Dimitrova,Ekaterina Dimotrova,edimitrova,Ekaterina Dimitrov,reviewed by Ekaterina Dimitrova
```

File: contribulyze/contribulyze.aliases [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+Jonathan Ellis,jbellis,jebllis
+Joel Knighton,jknighton
+Jordan West,jrwest
+Joey Lynch,Joseph Lynch
+Josh McKenzie,jmckenzie,josh,JoshuaMcKenzie,Joshua McKenzie
+Jun Rao,junrao
+Marcus Eriksson,Marcuse,Marcus,Krummas
+Matthew Dennis,mdennis
+Michael Kjellman,mkjellman
+Mick Semb Wever,Michael Semb Wever,Mck SembWever,Michael SembWever
```
Comment (JeremiahDJordan) [inline contribulyze/contribulyze.aliases]:
```suggestion
Mick Semb Wever,Michael Semb Wever,Mck SembWever,Michael SembWever,Mck Semb Wever
```

**Comment by JeremiahDJordan** [comment]
Added some suggestions for the alias file based on looking through the reports.  Also suggested a fix for the committer code to fix it from breaking if the two sets of ASF data don't match.

File: contribulyze/contribulyze.aliases (hunk 8 above)
Comment (michaelsembwever) [reply contribulyze/contribulyze.aliases:24]: these `reviewed by` errors are getting caused by something else. see @azotcsit's comment above https://github.com/apache/cassandra-builds/pull/54#discussion_r711124013

File: contribulyze/contribulyze.aliases (hunk 7 above)
Comment (michaelsembwever) [reply contribulyze/contribulyze.aliases:14]: these `reviewed by` errors are getting caused by something else. see @azotcsit's comment above https://github.com/apache/cassandra-builds/pull/54#discussion_r711124013

File: contribulyze/contribulyze.aliases (hunk 5 above)
Comment (michaelsembwever) [reply contribulyze/contribulyze.aliases:2]: these `reviewed by` errors are getting caused by something else. see @azotcsit's comment above https://github.com/apache/cassandra-builds/pull/54#discussion_r711124013

**Comment by michaelsembwever** [comment]
Working again after @JeremiahDJordan's fixes 💪 
https://ci-cassandra.apache.org/view/misc/job/contribulyze/86/

File: contribulyze/contribulyze.aliases [hunk 10]
```diff
@@ -0,0 +1,69 @@
+Aaron Morton,amorton
+Adam Holmberg,aholmber,Adam Holberg
+Adrian Cole,Adrain Cole
+Aleksandr Sorokoumov,Aleksandr Soromoukov
+Aleksey Yeschenko,iamaleksey,Aleksey,ayeschenko,Aleksey Yeshchenko,iamaleskey
+Aleksei Zotov,Alexey Zotov,Aleksei Zoto,azotcsit
```
Comment (azotcsit) [inline contribulyze/contribulyze.aliases]: One more variation has been added: Alexei Zotov

https://github.com/apache/cassandra/commit/243d220854895b968e2f169d3650c27d83567bb3

File: contribulyze/contribulyze.aliases [hunk 11]
```diff
@@ -0,0 +1,69 @@
+Aaron Morton,amorton
+Adam Holmberg,aholmber,Adam Holberg
+Adrian Cole,Adrain Cole
+Aleksandr Sorokoumov,Aleksandr Soromoukov
+Aleksey Yeschenko,iamaleksey,Aleksey,ayeschenko,Aleksey Yeshchenko,iamaleskey
+Aleksei Zotov,Alexey Zotov,Aleksei Zoto,azotcsit,Alexei Zotov
```
Comment (azotcsit) [inline contribulyze/contribulyze.aliases]: Got one more variant :smile: 

Alexander Zotov

https://github.com/apache/cassandra/commit/5fdadb25f95099b8945d9d9ee11d3e380d3867f4

**Comment by michaelsembwever** [comment]
Created https://issues.apache.org/jira/browse/CASSANDRA-17185 for final acceptance before merging.

**Comment by michaelsembwever** [comment]
manually merged with https://github.com/apache/cassandra-builds/commit/482c82bec11ab6cb13bb21bb1b52da200b548528 and https://github.com/apache/cassandra-builds/commit/e761fabe4e933d12b8402f0efcd14b14265ea435

**Pull request closed without merging**